src/support/small_set.h

```diff
@@ -0,0 +1,298 @@
+  FlexibleSet flexible;
+
+  bool usingFixed() const {
+    // If the flexible storage contains something, then we are using it.
+    // Otherwise we use the fixed storage. Note that if we grow and shrink then
+    // we will stay in flexible mode until we reach a size of zero, at which
+    // point we return to fixed mode. This is intentional, to avoid a lot of
+    // movement in switching between fixed and flexible mode.
+    return flexible.empty();
+  }
+
+public:
+  using value_type = T;
+  using key_type = T;
+  using reference = T&;
+  using const_reference = const T&;
+  using set_type = FlexibleSet;
+  using size_type = size_t;
+
+  SmallSetBase() {}
+  SmallSetBase(std::initializer_list<T> init) {
+    for (T item : init) {
+      insert(item);
+    }
+  }
+
+  void insert(const T& x) {
+    if (usingFixed()) {
+      if (count(x)) {
+        // The item already exists, so there is nothing to do.
+        return;
+      }
+      // We must add a new item.
+      if (usedFixed < N) {
+        // Room remains in the fixed storage.
+        fixed[usedFixed++] = x;
+      } else {
+        // No fixed storage remains. Switch to flexible.
+        assert(usedFixed == N);
+        assert(flexible.empty());
+        for (size_t i = 0; i < usedFixed; i++) {
+          flexible.insert(fixed[i]);
+        }
+        flexible.insert(x);
+        assert(!usingFixed());
+        usedFixed = 0;
+      }
+    } else {
+      flexible.insert(x);
+    }
+  }
+
+  void erase(const T& x) {
+    if (usingFixed()) {
+      for (size_t i = 0; i < usedFixed; i++) {
+        if (fixed[i] == x) {
+          // We found the item; erase it by moving the final item to replace it
+          // and truncating the size.
+          usedFixed--;
+          fixed[i] = fixed[usedFixed];
+          return;
+        }
+      }
+    } else {
+      flexible.erase(x);
+    }
+  }
+
+  size_t count(const T& x) {
+    if (usingFixed()) {
+      // Do a linear search.
+      for (size_t i = 0; i < usedFixed; i++) {
+        if (fixed[i] == x) {
+          return 1;
+        }
+      }
+      return 0;
+    } else {
+      return flexible.count(x);
+    }
+  }
+
+  size_t size() const {
+    if (usingFixed()) {
+      return usedFixed;
+    } else {
+      return flexible.size();
+    }
+  }
+
+  bool empty() const { return size() == 0; }
+
+  void clear() {
+    usedFixed = 0;
+    flexible.clear();
+  }
+
+  bool operator==(const SmallSetBase<T, N, FlexibleSet>& other) const {
+    if (usingFixed()) {
+      if (usedFixed != other.usedFixed) {
+        return false;
+      }
+      for (size_t i = 0; i < usedFixed; i++) {
+        bool found = false;
+        for (size_t j = 0; j < usedFixed; j++) {
+          if (fixed[i] == other.fixed[j]) {
+            found = true;
+            break;
+          }
+        }
+        if (!found) {
+          return false;
+        }
+      }
+      return true;
+    }
+    return flexible == other.flexible;
```

Comment:
Do you think this will be as efficient? I'm not sure myself...

Reply:
Hmm, yeah, it seems calling `count` can involve duplicate `usingFixed` checking.. Nevermind then.

Reply:
I don't feel strongly, but two downsides:

1. This breaks the code pattern in all the other functions, which is consistently "if usingFixed .. else ..".
2. This has repeated work in principle, as `flexible == other.flexible` will check the size once more. Perhaps the optimizer fixes that up though. But when I read it, the repetition seems odd to me personally.

Reply:
I don't think either the original code or any of the suggestions is quite correct: it is possible right now for equivalent sets to differ in whether they use `fixed` because removing elements never transitions back to using `fixed`. It would be good to have a test for this case.

In general I think this would be much more readable using `std::all_of` where possible instead of these small loops that spread the logic out over many lines.

Reply:
> it is possible right now for equivalent sets to differ in whether they use fixed because removing elements never transitions back to using fixed.

Good point! Haven't caught that. But then wouldn't my original suggestion work?
```cpp
   if (size() != other.size())
      return false;
    return std::all_of(begin(), end(), [&other](const T &x) { return other.count(x); });
```

Reply:
Ah, good catch! We do transition back to fixed mode when we reach a size of zero, and so you are right that a size S that is `0 < S <= N` may be either fixed *or* flexible mode, if we became flexible and shrank back down... fixed and added testing.

Reply:
The new code still uses `count()` and looks more verbose... Is that more efficient than https://github.com/WebAssembly/binaryen/pull/4188#discussion_r718041408?

Reply:
@aheejin I think you're right and your suggestion would work. In the code I just pushed though I implemented it to loop over the fixed set, if there is one, as that one would generally be smaller, which I hope is faster?

Reply:
Oh wait, they are the same size here... so I'm not sure what's best.

Reply:
I am not sure, but in the case that both are flexible, I'd prefer to leave things to the stdlib, as maybe it has some special optimizations for huge sets. So maybe this could be "if one is flexible, then do all_of, otherwise do flexible == other.flexible" - ?

Reply:
¯\_(ツ)_/¯ I'm fine either way.